File: docs/examples/sample.rb
# Sample Ruby file demonstrating module/class structure
require 'json'
require 'net/http'
require_relative 'utils/helpers'
require_relative 'models/base'

module Greeter
  VERSION = '1.0.0'.freeze

  module Formatter
    def format_name(name)
      name.strip.capitalize
    end
  end
end

class Animal
  include Greeter::Formatter
  extend Comparable

  attr_accessor :name, :age
  attr_reader :species
  attr_writer :habitat

  def initialize(name, species, age = 0)
    @name = name
    @species = species
    @age = age
    @habitat = 'unknown'
  end

  def speak
    "#{@name} says hello"
  end

  def to_s
    "#{@species}: #{@name} (age #{@age})"
  end

  def to_json(*args)
    { name: @name, species: @species, age: @age }.to_json
  end

  private

  def internal_id
    "#{@species}-#{@name.downcase}"
  end

  def log_action(action)
    puts "[#{Time.now}] #{action} for #{@name}"
  end

  protected

  def compare_age(other)
    @age <=> other.age
  end
end

class Dog < Animal
  include Greeter

  attr_accessor :breed

  def initialize(name, breed, age = 0)
    super(name, 'Canis lupus familiaris', age)
    @breed = breed
  end

  def speak
    "#{@name} barks: Woof!"
  end

  def fetch(item)
    "#{@name} fetches the #{item}!"
  end

  private

  def validate_breed
    raise ArgumentError, "Breed cannot be empty" if @breed.nil? || @breed.empty?
  end
end

module Serializable
  def self.included(base)
    base.extend(ClassMethods)
  end

  module ClassMethods
    def from_json(json_str)
      data = JSON.parse(json_str)
      new(data['name'], data['species'], data['age'])
    end
  end

  def serialize
    to_json
  end
end

File: docs/examples/sample.lua
-- sample.lua: A Lua module demonstrating common patterns

local json = require("json")
local utils = require("utils.helpers")
local http = require("socket.http")

-- Class-like pattern
Animal = {}
Animal.__index = Animal

function Animal:new(name, sound)
  local self = setmetatable({}, Animal)
  self.name = name
  self.sound = sound
  return self
end

function Animal:speak()
  return self.name .. " says " .. self.sound
end

-- Subclass
Dog = setmetatable({}, { __index = Animal })

function Dog:new(name)
  local self = Animal.new(self, name, "Woof")
  return setmetatable(self, { __index = Dog })
end

function Dog:fetch(item)
  return self.name .. " fetches the " .. item
end

-- Global functions
function greet(who)
  return "Hello, " .. tostring(who) .. "!"
end

function deepCopy(orig)
  local copy = {}
  for k, v in pairs(orig) do
    if type(v) == "table" then
      copy[k] = deepCopy(v)
    else
      copy[k] = v
    end
  end
  return copy
end

-- Local helper
local function clamp(value, min, max)
  if value < min then return min end
  if value > max then return max end
  return value
end

local function lerp(a, b, t)
  t = clamp(t, 0, 1)
  return a + (b - a) * t
end

-- Configuration table
local config = {
  version = "1.0.0",
  debug = false,
  maxRetries = 3,
  endpoints = {
    api = "https://api.example.com",
    auth = "https://auth.example.com",
  },
}

-- Return module table
return {
  Animal = Animal,
  Dog = Dog,
  greet = greet,
  deepCopy = deepCopy,
  lerp = lerp,
  config = config,
}
